### PR TITLE
Fix moving head tests

### PR DESCRIPTION
**Pull request by xian**
Oops, [broken tests](https://app.circleci.com/pipelines/github/baaahs/sparklemotion/2796/workflows/9fa7c35d-cc3e-4303-8fcb-a0583b680dd1/jobs/2447).

### CODE REVIEW

**Comment by xian** [comment]
There's still a failing test, which I think points out an actual bug.

If a shader declares `struct MovingHeadParams` with a subset of fields (e.g. without `prism`) then Sparkle Motion will emit it namespaced (e.g. `p0_untitledShader_MovingHeadParams`), which we don't seem to be doing. Not quite sure why yet.

**Comment by kcking** [comment]
Let me see if I can describe the failure:

Structs defined by the user are namespaced to prevent conflicting with structs defined by the shader generation layer (ie. `MovingHeadParams`).

If the user defines a struct with the exact same name and fields as a generated struct, then the namespacing is skipped, and the generated GLSL just has the generated non-namespaced struct.

If the user defines a struct with the same name, but different fields, a namespaced version of the struct is still generated, but instantiations of the struct are not converted to the namespaced version. i.e. `MovingHeadParams(<fields>)` will not be converted to `p0_untitledShader_MovingHeadParams(<fields>)`. This means it's not possible to refer to the namespaced definition of the struct, and it's essentially acting as documentation.

===

Perhaps this is fine behavior? The good thing is that existing shaders don't need to be updated when a field is added, which is what I was going for with the prism PR.

AFAIU, if we were to namespace all instantiations of `MovingHeadParams` by the user, then it wouldn't be possible to refer to the generated non-namespaced version of `MovingHeadParams` when, for example, returning it from `main`.

So I would propose either adding the missing fields to the test (removing the namespaced MovingHeadParams from the generated GLSL), or keeping the fields missing but updating the spec to the current behavior.

@xian WDYT?

**Comment by xian** [comment]
> Structs defined by the user are namespaced to prevent conflicting with structs defined by the shader generation layer (ie. `MovingHeadParams`).

Yep, and with other shaders that might be smooshed in with it.

> If the user defines a struct with the exact same name and fields as a generated struct, then the namespacing is skipped, and the generated GLSL just has the generated non-namespaced struct.
> 
> If the user defines a struct with the same name, but different fields, a namespaced version of the struct is still generated, but instantiations of the struct are not converted to the namespaced version. i.e. `MovingHeadParams(<fields>)` will not be converted to `p0_untitledShader_MovingHeadParams(<fields>)`. This means it's not possible to refer to the namespaced definition of the struct, and it's essentially acting as documentation.

Right, I _think_ there is some special behavior about identical strucgts and subsets of structs but I can't remember exactly what it is, and I'm not 100% certain I ever implemented that idea.

> Perhaps this is fine behavior? The good thing is that existing shaders don't need to be updated when a field is added, which is what I was going for with the prism PR.

Yep, the goal was that additional struct members can be added without breaking old shaders.

> AFAIU, if we were to namespace all instantiations of `MovingHeadParams` by the user, then it wouldn't be possible to refer to the generated non-namespaced version of `MovingHeadParams` when, for example, returning it from `main`.
> 
> So I would propose either adding the missing fields to the test (removing the namespaced MovingHeadParams from the generated GLSL), or keeping the fields missing but updating the spec to the current behavior.

Does it still work to intermix shaders that do and don't refer to `prism`? If so, maybe I'm misunderstanding what I'm seeing in the spec, but it looks to me like that spec wouldn't compile?

**Comment by kcking** [comment]
I see your point.
```
MovingHeadParams p0_untitledShader_main() {
    p0_untitledShader_AnotherStruct b;
    return MovingHeadParams(p0_untitledShader_a.first, p0_untitledShader_a.second, 0., 0.);
}
```
will fail to compile because `MovingHeadParams` has 6 fields.

However,
```
MovingHeadParams p0_untitledShader_main() {
    p0_untitledShader_AnotherStruct b;
    return p0_untitledShader_MovingHeadParams(p0_untitledShader_a.first, p0_untitledShader_a.second, 0., 0.);
}
```
will fail to compile because these are not the same struct.

The spec uses a "return from main" syntax
```
MovindHeadParams main() {
    return MovingHeadParams(...);
}
```
While the actual shader template uses an "out param" syntax
```
// @param params moving-head-params
void main(out MovingHeadParams params) {
            params.pan = 0.;
            params.tilt = .5;
            params.colorWheel = 0.;
            params.dimmer = 1.;
            params.prism = false;
            params.prismRotation = 0.;
}
```

Interestingly, the latter allows backwards compatibility. An old shader will just not set `params.prism`, and it will be left at the default value. WDYT about changing the spec to match [the template](https://github.com/baaahs/sparklemotion/blob/81d3c3e775016ce119d6bbda9d58108ac18b2689/src/commonMain/kotlin/baaahs/gl/shader/type/MoverShader.kt#L39-L47)? I'm happy to do that

_I wonder if backwards compatibility was the reason the template was changed to an outparam, but the spec just wasn't updated?_

File: src/commonTest/kotlin/baaahs/TestMovingHeadAdapter.kt [hunk 1]
```diff
@@ -25,6 +25,9 @@ class TestMovingHeadAdapter(
     override val tiltRange: ClosedRange<Float> = toRadians(-135f)..toRadians(135f),
     override val tiltMotorSpeed: Float = 1f,
 
+    override val prismChannel: Dmx.Channel = TestChannel(5),
+    override val prismRotationChannel: Dmx.Channel = TestChannel(6),
```
Comment (kcking) [inline src/commonTest/kotlin/baaahs/TestMovingHeadAdapter.kt:29]: this conflicts with the shutterChannel number below, maybe bump that one up to 7 as well

**Comment by xian** [comment]
I'm putting in a TODO :-(

**Pull request merged**